Default auth scheme handling (#30300)
dotnet/docs: docs/core/compatibility/aspnet-core/7.0/default-authentication-scheme.md

---
title: "Breaking change: Default authentication scheme"
description: Learn about the breaking change in ASP.NET Core 7.0 where a singly registered authentication scheme will cause it to be used as the default.
ms.date: 07/20/2022
---
# Default authentication scheme

Starting in .NET 7, we introduced new behavior in the authentication area in ASP.NET Core.

Previously, users were required to set the default authentication scheme, which is used by authentication and authorization handlers, in the `AddAuthentication` call:

```csharp
builder.Services.AddAuthentication("MyDefaultScheme");
```

> Moving forward, when a *single* authentication scheme is registered, that scheme will be treated as the default scheme. For example, "myscheme" will be treated as the default scheme in the following code.
> 
> ```csharp
> builder.Services.AddAuthentication().AddOAuth("myscheme");
> ```

Reviewer: ```suggestion
Moving forward, when a *single* authentication scheme is registered, that scheme will be treated as the default scheme. For example, "MyDefaultScheme" will be treated as the default scheme in the following code.

```csharp
builder.Services.AddAuthentication().AddOAuth("MyDefaultScheme");
```
```

This change might expose unintended behavior changes in applications, such as authentication options being validated earlier than expected.

## Version introduced

ASP.NET Core 7.0 Preview 7

## Previous behavior

Previously, when users did not provide a default scheme in the `AddAuthentication` call, no default scheme was set.

```csharp
builder.Services.AddAuthentication().AddCookie();
```

This impacted the behavior of authentication handlers in the application layer.

## New behavior

> Starting in ASP.NET Core 7.0, if (and only if) a single scheme is registered in an application, that scheme is treated as the default. In the following code, the `CookieDefaults.AuthenticationScheme` is treated as the default scheme.

Reviewer: ```suggestion
Starting in ASP.NET Core 7.0, when a single scheme is registered in an application, that scheme is treated as the default. In the following code, the `CookieDefaults.AuthenticationScheme` is treated as the default scheme.
```

Author: I would like to keep this as-is to emphasize that it's in that specific case only.

Reviewer 2: Agreed. I included this language specifically because there was confusion in the announcement's discussion about whether this policy applied if more than one scheme was registered (it doesn't).
